pkg/gcptarget/library.go

```diff
@@ -0,0 +1,71 @@
+package gcptarget
+
+import "text/template"
+
+const libraryTemplateSrc = `package target
+
+matching_constraints[constraint] {
+	asset := input.review
+	constraint := {{.ConstraintsRoot}}[_][_]
+	spec := get_default(constraint, "spec", {})
+	match := get_default(spec, "match", {})
+
+	# Default matcher behavior is to match everything.
+	target := get_default(match, "target", ["organization/*"])
+	# TODO: Retire the "gcp" wrapper.
+	# See https://github.com/forseti-security/config-validator/issues/42
+	gcp := get_default(match, "gcp", {})
+	gcp_target := get_default(gcp, "target", target)
+	re_match(gcp_target[_], asset.ancestry_path)
+	exclude := get_default(match, "exclude", [])
+	gcp_exclude := get_default(gcp, "exclude", exclude)
+	exclusion_match := {asset.ancestry_path | re_match(gcp_exclude[_], asset.ancestry_path)}
+	count(exclusion_match) == 0
+}
+
+# CAI Resource Types
+matching_reviews_and_constraints[[review, constraint]] {
+  review = {{.DataRoot}}[_]
+  matching_constraints[constraint] with input as {"review": review}
+}
+
+autoreject_review[rejection] {
+	false
+  rejection := {
+		"msg": "should not reach this", 
+	}
+}
+
+########
+# Util #
+########
+# get_default returns the value of an object's field or the provided default value.
+# It avoids creating an undefined state when trying to access an object attribute that does
+# not exist
+get_default(object, field, _default) = output {
+  has_field(object, field)
+  output = object[field]
+}
+
+get_default(object, field, _default) = output {
+  has_field(object, field) == false
+  output = _default
+}
+
+# has_field returns whether an object has a field
+has_field(object, field) = true {
+  object[field]
+}
+# False is a tricky special case, as false responses would create an undefined document unless
+# they are explicitly tested for
+has_field(object, field) = true {
+  object[field] == false
+}
+has_field(object, field) = false {
+  not object[field]
+  not object[field] == false
+}
+
+`
```

Comment:
any Rego tests for this source code?

Reply:
just being tested via the unit test stuff

Reply:
ack

Reply:
ack